versions/1.1/SPEC.md

```diff
@@ -0,0 +1,5775 @@
+
+In WDL *all* types represent immutable values.
+
+- Even a `File` represent a logical "snapshot" of the file at the time when the value was created.
+- It's impossible for a task to change an upstream value which has been provided as an input: even if it makes changes to its local copy the original value is unaffected.
+
+#### Primitive Types
+
+The following primitive types exist in WDL:
+
+* A `Boolean` represents a value of `true` or `false`.
+* An `Int` represents a signed integer in the range \[-2^63, 2^63).
+* A `Float` represents a finite 64-bit IEEE-754 floating point number.
+* A `String` represents a unicode character string following the format described [above](#strings).
+* A `File` represents a file (or file-like object).
+  * A `File` declaration can have a string value indicating a relative or absolute path on the local file system.
+  * Within a WDL file, literal values for files may only be local (relative or absolute) paths.
+  * An execution engine may support other ways to specify [`File` inputs (e.g. as URIs)](#input-and-output-formats), but prior to task execution it must [localize inputs](#task-input-localization) so that the runtime value of a `File` variable is a local path.
+
+Examples:
+
+```wdl
+Boolean b = true 
+Int i = 0
+Float f = 27.3
+String s = "hello, world"
+File f = "path/to/file"
+```
+
+#### Compound Types
+
+A compound type is one that contains nested types, i.e. it is *parameterized* by other types. The following compound types can be constructed. In the examples below `P` represents any of the primitive types above, and `X` and `Y` represent any valid type (including nested compound types).
+
+##### Array[X]
+
+An `Array` represents an ordered list of elements that are all of the same type.
+
+An array value can be declared with comma-separated values in brackets (`[]`), and a specific (zero-based) index of an `Array` can be accessed by placing the index in brackets after the declaration name. Accessing a non-existent index of an `Array` results in an error.
+
+```wdl
+Array[File] files = ["/path/to/file1", "/path/to/file2"]
+File f = files[0]
+
+Array[Int] empty = []
+empty[0]  # error!
+```
+
+An `Array` may be empty, unless it is declared using the non-empty postfix quantifier `+`, in which case it must contain at least one element. Note that the `+` and `?` postfix quantifiers can be combined to declare an array that is either undefined or non-empty. Attempting to assign an empty array literal to a non-empty `Array` declaration results in an error. Otherwise, the non-empty assertion is only checked at runtime: binding an empty array to an `Array[T]+` input or function argument is a runtime error. 
```

Comment:
Is there a more concrete example of when you'd want to declare an undefined array?  Or a non-empty one?  Like a one-liner explaining this?

Reply:
I tried to make this more clear. Please read the revised version and see if it make sense.